### ISSUE TITLE
Duplicate readlists and missed books in readlists on scan with StoryArc in ComicInfo.xml.

### ISSUE THREAD

**Issue by Shakahron** [gotson/komga#1317]
### Steps to reproduce

Scan library with many readlists set with StoryArc in embedded ComicInfo file.

### Expected behavior

There shouldn't be duplate readlists and all books should be added to readlists from a normal scan.

### Actual behavior

I have a library full of oneshots with many of them being in a readlist so I don't know if this is an issue for books in series or not. I noticed that occasionally  a duplicate readlist with a single book in it will be created and sometimes books won't be added to a readlist unless you specifically scan those books for metadata.

### Komga version

1.8.2

### Operating system

Windows 10

### Installation method

from download.komga.org

### Acknowledgements

- [X] I have searched the existing issues (open AND closed) and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I have checked the [FAQ](https://komga.org/docs/faq).
- [X] I have updated the app to the latest version.
- [X] I will fill out all of the requested information in this form.

**Comment by gotson** [comment]
I don't quite understand what the problem is, could you explain more or provide some screenshots maybe?

**Comment by Shakahron** [comment]
I'll try and explain more. I have a library made up entirely of oneshots, so the oneshots folder is set as the library folder. There are a lot of readlists set with the <StoryArc> and <StoryArcNumber> tags in ComicInfo.xml's in this library, so going through them individually would be a time consuming process. 

When I scan this library, not all of the books are going into the same readlist. Some are duplicated, some are split off into a seperate readlist and sometimes books are meant to go into readlists aren't found unless you refresh metadata. I've blurred the screenshots as they are adult:

This one shows 2 readlists where 1 readlist is already complete with 4 books, but a duplicate was made with 1 book.
![firefox_giVbW7jKXe](https://github.com/gotson/komga/assets/106402227/720afb66-ad62-4626-8c91-0e332904fc56)

This one show 2 readlists with 1 book, where there should be 1 readlist with 2 books.
![firefox_mMPpARse0e](https://github.com/gotson/komga/assets/106402227/85e59caa-9313-4df3-9d28-8c1e937fea1f)

Here there are 6 books in this readlist where there should be 7. When I refresh the library metadata, sometimes the book is added, but sometime I need to go and metadata for the book itself.
![firefox_nekVUwP30E](https://github.com/gotson/komga/assets/106402227/dc23fa20-5ec6-4a52-ad2e-2af124c4bdaf)

I hope that makes it clearer what the problem was.

**Comment by gotson** [comment]
can you retrieve the API response for `GET /api/v1/readlists` ? I'd need to have a look at the raw data in there.

**Comment by gotson** [comment]
Side question, how many threads did you configure for background tasks?

**Comment by Shakahron** [comment]
I configured 8 threads for background tasks

**Comment by gotson** [comment]
> I configured 8 threads for background tasks

can you set it to 1 and try if you have the problem ? I surmise that processing in parallel may create duplicate read lists.

**Comment by Shakahron** [comment]
I'll try that now. by the for your first question, I'm not sure how to do that, can you point me in a general direction to get that for you?

**Comment by Shakahron** [comment]
I figured it out. I'll get the API response fter I finish testing 1 thread.

**Comment by Shakahron** [comment]
The library hasn't finished scanning yet but 100% confident in say that setting background tasks to 1 thread has completely fixed the issues. Do you still want the api response for readlists?

**Comment by gotson** [comment]
I have been testing this on my side, even though i could not reproduce the duplicate read list (probably because my database / cpu are too fast on my dev machine), i can see that it could happen. I could definitely reproduce the issue where a read list did not contain all the items it was supposed to contain.

It's a multi-threading issue, there should be some resource locking when creating/updating readlists (and collections too), which is missing.

I am adding that now.

**Comment by gotson** [comment]
The fix won't remove the duplicates though, you will need to do the cleanup yourself.

**Comment by Shakahron** [comment]
Do you mean the fix won't remove duplicates already in the library? Or are duplicates just something that can't be prevented with more than one thread?

**Comment by gotson** [comment]
> Do you mean the fix won't remove duplicates already in the library? Or are duplicates just something that can't be prevented with more than one thread?

the fix will not remove the existing duplicates, you will have to delete the duplicates yourselves.

the fix will prevent new duplicates (hopefully!!)

**Comment by Shakahron** [comment]
That's not a problem at all, thanks so much for taking the time on this.
